There's some weird interaction between `iscomplexobj` and `assert_equal` (which internally calls `iscomplexobj`) that is causing the observed test failures, e.g.,
```

self = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch object at 0x7f8910047ba8>
    def test_interface(self):
    
        class MyArray(object):
            def __array_function__(self, func, types, args, kwargs):
                return (self, func, types, args, kwargs)
    
        original = MyArray()
        (obj, func, types, args, kwargs) = dispatched_one_arg(original)
        assert_(obj is original)
        assert_(func is dispatched_one_arg)
        assert_equal(set(types), {MyArray})
>       assert_equal(args, (original,))
MyArray    = <class 'numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray'>
args       = (<numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>,)
func       = <function dispatched_one_arg at 0x7f8913644e18>
kwargs     = {}
obj        = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>
original   = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>
self       = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch object at 0x7f8910047ba8>
types      = [<class 'numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray'>]
../builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_overrides.py:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../builds/venv/lib/python3.6/site-packages/numpy/core/overrides.py:151: in public_api
    implementation, public_api, relevant_args, args, kwargs)
../builds/venv/lib/python3.6/site-packages/numpy/core/overrides.py:96: in array_function_implementation_or_override
    return implementation(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = 5
    @array_function_dispatch(_iscomplexobj_dispatcher)
    def iscomplexobj(x):
        """
        Check for a complex type or an array of complex numbers.
    
        The type of the input is checked, not the value. Even if the input
        has an imaginary part equal to zero, `iscomplexobj` evaluates to True.
    
        Parameters
        ----------
        x : any
            The input can be of any type and shape.
    
        Returns
        -------
        iscomplexobj : bool
            The return value, True if `x` is of a complex type or has at least
            one complex element.
    
        See Also
        --------
        isrealobj, iscomplex
    
        Examples
        --------
        >>> np.iscomplexobj(1)
        False
        >>> np.iscomplexobj(1+0j)
        True
        >>> np.iscomplexobj([3, 1+0j, True])
        True
    
        """
        try:
>           dtype = x.dtype
E           RecursionError: maximum recursion depth exceeded
x          = 5
../builds/venv/lib/python3.6/site-packages/numpy/lib/type_check.py:311: RecursionError
```

--- a/numpy/core/tests/test_overrides.py
+++ b/numpy/core/tests/test_overrides.py
@@ -16,8 +16,8 @@ def _get_overloaded_args(relevant_args):
     return args
 
 
-def _return_self(self, *args, **kwargs):
-    return self
+def _return_not_implemented(self, *args, **kwargs):
+    return NotImplemented
 
 
 class TestGetOverloadedTypesAndArgs(object):
@@ -45,7 +45,7 @@ class TestGetOverloadedTypesAndArgs(object):
     def test_ndarray_subclasses(self):
 
         class OverrideSub(np.ndarray):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class NoOverrideSub(np.ndarray):
             pass
@@ -70,7 +70,7 @@ class TestGetOverloadedTypesAndArgs(object):
     def test_ndarray_and_duck_array(self):
 
         class Other(object):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         array = np.array(1)
         other = Other()
@@ -86,10 +86,10 @@ class TestGetOverloadedTypesAndArgs(object):
     def test_ndarray_subclass_and_duck_array(self):
 
         class OverrideSub(np.ndarray):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class Other(object):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         array = np.array(1)
         subarray = np.array(1).view(OverrideSub)
@@ -103,16 +103,16 @@ class TestGetOverloadedTypesAndArgs(object):
     def test_many_duck_arrays(self):
 
         class A(object):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class B(A):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class C(A):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class D(object):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         a = A()
         b = B()
@@ -135,7 +135,7 @@ class TestNDArrayArrayFunction(object):
     def test_method(self):
 
         class SubOverride(np.ndarray):
-            __array_function__ = _return_self
+            __array_function__ = _return_not_implemented
 
         class NoOverrideSub(np.ndarray):
             pass
@@ -187,7 +187,8 @@ class TestArrayFunctionDispatch(object):
         assert_(obj is original)
         assert_(func is dispatched_one_arg)
         assert_equal(set(types), {MyArray})
-        assert_equal(args, (original,))
+        # assert_equal uses the overloaded np.iscomplexobj() internally
+        assert_(args == (original,))
         assert_equal(kwargs, {})
 
     def test_not_implemented(self):

--- a/numpy/lib/twodim_base.py
+++ b/numpy/lib/twodim_base.py
@@ -34,11 +34,11 @@ def _min_int(low, high):
     return int64
 
 
-def _fliplr_dispatcher(m):
+def _flip_dispatcher(m):
     return (m,)
 
 
-@array_function_dispatch(_fliplr_dispatcher)
+@array_function_dispatch(_flip_dispatcher)
 def fliplr(m):
     """
     Flip array in the left/right direction.
@@ -89,11 +89,7 @@ def fliplr(m):
     return m[:, ::-1]
 
 
-def _flipud_dispatcher(m):
-    return (m,)
-
-
-@array_function_dispatch(_flipud_dispatcher)
+@array_function_dispatch(_flip_dispatcher)
 def flipud(m):
     """
     Flip array in the up/down direction.
@@ -281,11 +277,7 @@ def diag(v, k=0):
         raise ValueError("Input must be 1- or 2-d.")
 
 
-def _diagflat_dispatcher(v, k=None):
-    return (v,)
-
-
-@array_function_dispatch(_diagflat_dispatcher)
+@array_function_dispatch(_diag_dispatcher)
 def diagflat(v, k=0):
     """
     Create a two-dimensional array with the flattened input as a diagonal.
@@ -437,11 +429,7 @@ def tril(m, k=0):
     return where(mask, m, zeros(1, m.dtype))
 
 
-def _triu_dispatcher(m, k=None):
-    return (m,)
-
-
-@array_function_dispatch(_triu_dispatcher)
+@array_function_dispatch(_tril_dispatcher)
 def triu(m, k=0):
     """
     Upper triangle of an array.
@@ -969,11 +957,7 @@ def triu_indices(n, k=0, m=None):
     return nonzero(~tri(n, m, k=k-1, dtype=bool))
 
 
-def _triu_indices_form_dispatcher(arr, k=None):
-    return (arr,)
-
-
-@array_function_dispatch(_triu_indices_form_dispatcher)
+@array_function_dispatch(_tril_indices_form_dispatcher)
 def triu_indices_from(arr, k=0):
     """
     Return the indices for the upper-triangle of arr.

--- a/numpy/lib/ufunclike.py
+++ b/numpy/lib/ufunclike.py
@@ -38,12 +38,28 @@ def _deprecate_out_named_y(f):
     return func
 
 
-def _fix_dispatcher(x, out=None, **kwargs):
+def _fix_out_named_y(f):
+    """
+    Allow the out argument to be passed as the name `y` (deprecated)
+
+    This decorator should only be used if _deprecate_out_named_y is used on
+    a corresponding dispatcher fucntion.
+    """
+    @functools.wraps(f)
+    def func(x, out=None, **kwargs):
+        if 'y' in kwargs:
+            # we already did error checking in _deprecate_out_named_y
+            out = kwargs.pop('y')
+        return f(x, out=out, **kwargs)
+
+
+@_deprecate_out_named_y
+def _dispatcher(x, out=None):
     return (x, out)
 
 
-@array_function_dispatch(_fix_dispatcher, verify=False)
-@_deprecate_out_named_y
+@array_function_dispatch(_dispatcher, verify=False)
+@_fix_out_named_y
 def fix(x, out=None):
     """
     Round to nearest integer towards zero.
@@ -89,12 +105,8 @@ def fix(x, out=None):
     return res
 
 
-def _isposinf_dispatcher(x, out=None, **kwargs):
-    return (x, out)
-
-
-@array_function_dispatch(_isposinf_dispatcher, verify=False)
-@_deprecate_out_named_y
+@array_function_dispatch(_dispatcher, verify=False)
+@_fix_out_named_y
 def isposinf(x, out=None):
     """
     Test element-wise for positive infinity, return result as bool array.
@@ -162,12 +174,8 @@ def isposinf(x, out=None):
         return nx.logical_and(is_inf, signbit, out)
 
 
-def _isneginf_dispatcher(x, out=None, **kwargs):
-    return (x, out)
-
-
-@array_function_dispatch(_isneginf_dispatcher, verify=False)
-@_deprecate_out_named_y
+@array_function_dispatch(_dispatcher, verify=False)
+@_fix_out_named_y
 def isneginf(x, out=None):
     """
     Test element-wise for negative infinity, return result as bool array.

--- a/numpy/testing/_private/utils.py
+++ b/numpy/testing/_private/utils.py
@@ -352,7 +352,7 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
     # XXX: catch ValueError for subclasses of ndarray where iscomplex fail
     try:
         usecomplex = iscomplexobj(actual) or iscomplexobj(desired)
-    except ValueError:
+    except (ValueError, TypeError):
         usecomplex = False
 
     if usecomplex:

